sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,791 @@
+    # "Methods of Solution of the Riccati Differential Equation"
+    # by D. R. Haaheim and F. M. Stein
+
+    # If no particular solutions are found, a general
+    # solution cannot be found
+    if len(part_sols) == 0:
+        return []
+
+    # In case of a single particular solution, the general
+    # solution can be found by solving a Bernoulli ODE
+    elif len(part_sols) == 1:
+        y1 = part_sols[0]
+        z = exp(Integral(2*y1, x)) * Integral(exp(-Integral(2*y1, x)) * a, x)
+        z = z.doit()
+        if a == 0 or z == 0:
+            return y1
+        return y1 + 1/z
+
+    # In case of 2 particular solutions, the general solution
+    # can be found by solving a separable equation
+    elif len(part_sols) == 2:
+        y1, y2 = part_sols
+        u = exp(Integral(y2 - y1, x)).doit()
+        if u == 1:
+            return y2
+        return (y2*u - y1)/(u - 1)
+
+    # In case of 3 particular solutions, a closed form
+    # of the general solution can be obtained directly
+    else:
+        y1, y2, y3 = part_sols[:3]
+        C1 = Dummy('C1')
+        return (C1 + 1)*y2*(y1 - y3)/(C1*y1 + y2 - (C1 + 1)*y3)
+
+
+def solve_riccati(fx, x, b0, b1, b2, gensol=False):
+    # Step 1 : Convert to Normal Form
+    a = -b0*b2 + b1**2/4 - b1.diff(x)/2 + 3*b2.diff(x)**2/(4*b2**2) + b1*b2.diff(x)/(2*b2) - \
+        b2.diff(x, 2)/(2*b2)
+    a_t = a.together()
+    num, den = [Poly(e, x, extension=True) for e in a_t.as_numer_denom()]
+    num, den = num.cancel(den, include=True)
+
+    # Step 2
+    presol = []
+
+    # Step 3 : a(x) is 0
+    if num == 0:
+        presol.append(1/(x + Dummy('C1')))
+
+    # Step 4 : a(x) is a non-zero constant
+    elif x not in num.free_symbols.union(den.free_symbols):
+        presol.extend([sqrt(a), -sqrt(a)])
+
+    # Step 5 : Find poles and valuation at infinity
+    poles = find_poles(den, x)
+    poles, muls = list(poles.keys()), list(poles.values())
+    val_inf = val_at_inf(num, den, x)
+
+    if len(poles):
+        # Check necessary conditions (outlined in the module docstring)
+        if not check_necessary_conds(val_inf, muls):
+            raise ValueError("Rational Solution doesn't exist")
+
+        # Step 6
+        # Construct c-vectors for each singular point
+        c = construct_c(num, den, x, poles, muls)
+
+        # Construct d vectors for each singular point
+        d = construct_d(num, den, x, val_inf)
+
+        # Step 7 : Iterate over all possible combinations and return solutions
+        # For each possible combination, generate an array of 0's and 1's
+        # where 0 means pick 1st choice and 1 means pick the second choice.
+        choices = product(range(2), repeat=len(poles) + 1)
+        m_ybars = set()
+        for choice in choices:
+            # Step 8 and 9 : Compute m and ybar
+            m_ybars.add(compute_m_ybar(x, poles, choice, c, d, -val_inf//2))
+
+        for m, ybar in sorted(m_ybars, key=lambda x: x[0]):
```

Comment:
I don't see how the sorting helps here if there is no early exit from this loop.

Reply:
I was just thinking of adding an early exit, but it might happen that we exit if there are 3 solutions in `presol`, but it might happen that 2 of them are redundant, so instead of spending some more time in computing the particular solutions, the code will end up computing the general solution from a single particular solution which might be a lot slower. What do you think should be done?

Reply:
I get the impression that the early exit is not usually beneficial since we don't usually get 3 rational particular solutions. If that's the case then I don't think we need to pursue this.

Reply:
There should be some comments in the code explaining that an early exit is at least theoretically possible.